Cleanup output for easy import into visualization tools

reidmv opened the pull request:
This commit rewrites the metrics querying scripts to format the output in a way that is easy to import into visualization tools like Graphite/Grafana. By necessity it switches the scripts from bash to ruby, so that hashes can be built and manipulated.

reidmv committed:
Make output of puppetdb metrics cleaner

Changes:

  * Change "puppet_server" to "puppetserver" when referring to the
    product and naming files. This is consistent with how puppetserver
    is referred to in product and documentation.
  * Refactor puppetserver metrics script to store all metrics for a
    timestamp in a single JSON file, even when multiple servers are
    queried
  * Refactor puppetdb metrics script to store all metrics for a
    timestamp in a single JSON file, even when multiple servers are
    queried
  * Standardize file timestamp format to use ISO 8601
  * Standardize filename to always be ${timestamp}.json
  * Move data out of puppetdb template and into manifest. Adheres to
    best practices for templates, which encourage keeping data and code
    out of templates as much as possible.

These changes are targeted at simplifying data upload into Graphite /
Grafana.

reidmv committed:
Add error catching between hosts

So that if gathering metrics from one host fails, it won't prevent
others from being collected and written to file.

reidmv committed:
Add query time to data

So that the total time it took to run the query for a given server is
available.

reidmv committed:
Simplify templating and write to multiple files

reidmv committed:
Add timestamp to data files

reidmv committed:
Embed metrics data in template

While it's not awesome to have a conditional in a template, there
doesn't seem to be an obviously more elegant way to do this right now.

reidmv committed:
Implement outputdir/host/timestamp.json structure

This commit updates the scripts to create one directory per server
metrics are being obtained from, and to place files for each server in
its dedicated directory.

The idea is to make it easier to manually browse records on the
filesystem using cd and ls, for very large sets of data.

reidmv committed:
Fix metrics_ensure=absent behavior

Make it so that if a metric is ensured absent, the outputdir is also
ensured absent

reidmv committed:
Add deprecated support for old puppet_server API

So that customers already using this module don't have to change their
own code to use the new version of the module.

npwalker commented on manifests/init.pp:
```diff
@@ -27,4 +27,10 @@
+  # Ensure remanants of cron jobs and files from older versions of this module
+  # are cleaned up.
+  pe_metric_curl_cron_jobs::pe_metric { 'puppet_server' :
+    ensure => 'absent',
```
Should be `metric_ensure`

```
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Pe_metric_curl_cron_jobs::Pe_metric[puppet_server]: has no parameter named 'ensure' at /etc/puppetlabs/code/environments/production/modules/pe_metric_curl_cron_jobs/manifests/init.pp:32 on node master201643-centos.puppetdebug.vlan
```

djjudas21 commented:
Works for me :+1:

reidmv commented:
@npwalker any update?

npwalker commented:
@reidmv I've asked Charlie to take a look at it then I think we can merge.

Sharpie commented:
👍 looks fine to me!

djjudas21 commented:
The cleanup script which runs nightly runs find `'/opt/puppetlabs/pe_metric_curl_cron_jobs' -type f -mtime 3 -delete -print` to delete old reports. The `-print` flag is causing the `find` command to produce output like the following on stdout - which is then emailed to us:

```
/opt/puppetlabs/pe_metric_curl_cron_jobs/puppetserver/127.0.0.1/20170304T025501Z.json
/opt/puppetlabs/pe_metric_curl_cron_jobs/puppetdb/127.0.0.1/20170304T025501Z.json
```

We don't really care what it deleted - so can we remove `-print` from that cronjob and save a few emails per night? Happy to send a PR myself but it might be easier to tack another commit onto this one.

reidmv committed:
Adhere to Unix philosophy "Rule of Silence"

The cleanup script which runs nightly runs `find [...] -delete -print`
to delete old reports. The -print flag causes the find command to
produce output like the following on stdout - which is then emailed to
users.

    /opt/puppetlabs/pe_metric_curl_cron_jobs/puppetserver/127.0.0.1/20170304T025501Z.json
    /opt/puppetlabs/pe_metric_curl_cron_jobs/puppetdb/127.0.0.1/20170304T025501Z.json

This is a normal recurring activity and does not need to be logged.
Sending emails for it violates the "Rule of Silence" in the Unix
philosophy. When a program has nothing surprising to say, it should say
nothing.

This commit aligns the retention cleanup cron job with the Unix
philosophy's rule of silence.
